src/iterator.cr

```diff
@@ -381,84 +379,74 @@ module Iterator(T)
   #     iter.next # => 'b'
   #     iter.next # => Iterator::Stop::INSTANCE
   #
-  # def flatten
-  #   Flatten(self, typeof(Flatten.element_type(self))).new(self)
-  # end
-
-  # private class Flatten(I, T1)
-  #   include Iterator(T1)
-
-  #   @iterator : I
-  #   @top : Bool
-  #   @to_rewind : Array(Proc(Nil))
-
-  #   # @generator : I
-
-  #   def initialize(@iterator)
-  #     @generator = @iterator
-  #     @top = true
-  #     @to_rewind = [] of Proc(Nil)
-  #   end
-
-  #   def next
-  #     value = @generator.next
-  #     if value.is_a?(Stop)
-  #       if @top
-  #         return stop
-  #       else
-  #         @generator = @iterator
-  #         @top = true
-  #         return self.next
-  #       end
-  #     end
-
-  #     flatten value
-  #   end
-
-  #   def make_rewinder(iter)
-  #     ->{
-  #       iter.rewind
-  #       # Return nil to disguise the individual iterator types
-  #       nil
-  #     }
-  #   end
-
-  #   def rewind
-  #     @iterator.rewind
-  #     @generator = @iterator
-  #     @top = true
-  #     @to_rewind.each &.call
-  #     @to_rewind.clear
-  #   end
-
-  #   def flatten(element)
-  #     case element
-  #     when Iterator
-  #       flat = element.flatten
-  #       @generator = flat
-  #       @to_rewind << make_rewinder flat
-  #       @top = false
-  #       self.next
-  #     when Iterable
-  #       flatten element.each
-  #     else
-  #       element
-  #     end
-  #   end
-
-  #   def self.element_type(element)
-  #     case element
-  #     when Stop
-  #       raise ""
-  #     when Iterator
-  #       element_type(element.next)
-  #     when Iterable
-  #       element_type(element.each)
-  #     else
-  #       element
-  #     end
-  #   end
-  # end
+  def flatten
+    Flatten(typeof(Flatten.iterator_type(self)), typeof(Flatten.element_type(self))).new(self)
+  end
+
+  private class Flatten(I, T)
+    include Iterator(T)
+
+    @iterator : I
+    @stopped : Array(I)
+    @generators : Array(I)
+
+    def initialize(@iterator)
+      @generators = [@iterator]
+      @stopped = [] of I
+    end
+
+    def next
+      case value = @generators.last.next
```

Comment:
I don't know, you can try to check if LLVM generates `tailcc` (or something like that, can't remember now) when doing `--emit llvm-ir` (though finding the function definition for this method might be tricky, it will be named something like `Iterator#next`)